File: app/services/attendance_job_labels.py
"""Resolve attendance job_type / reason_text markers to human-readable labels."""
from __future__ import annotations

import uuid
from typing import Dict, Iterable, List, Optional, Tuple, Union

from sqlalchemy.orm import Session

from ..models.models import Project

PREDEFINED_JOBS_DICT: Dict[str, str] = {
    "0": "No Project Assigned",
    "37": "Repairs",
    "47": "Shop",
    "53": "YPK Developments",
    "136": "Stat Holiday",
}


def parse_job_type_from_reason_text(reason_text: Optional[str]) -> Optional[str]:
    if not reason_text or not reason_text.startswith("JOB_TYPE:"):
        return None
    marker = reason_text.split("|", 1)[0]
    job_type = marker.replace("JOB_TYPE:", "", 1).strip()
    return job_type or None


def format_project_job_label(project: Union[Project, object]) -> str:
    name = (getattr(project, "name", None) or "").strip()
    code = (getattr(project, "code", None) or "").strip()
    if name and code:
        return f"{name} ({code})"
    return name or code or "Unknown"


def _is_valid_uuid(value: str) -> bool:
    try:
        uuid.UUID(str(value))
        return True
    except (ValueError, TypeError, AttributeError):
        return False


def collect_project_ids_from_job_types(job_types: Iterable[Optional[str]]) -> List[uuid.UUID]:
    ids: List[uuid.UUID] = []
    seen = set()
    for job_type in job_types:
        if not job_type or job_type in PREDEFINED_JOBS_DICT:
            continue
        if not _is_valid_uuid(job_type):
            continue
        uid = uuid.UUID(str(job_type))
        key = str(uid)
        if key in seen:
            continue
        seen.add(key)
        ids.append(uid)
    return ids


def load_projects_by_id(db: Session, project_ids: Iterable[uuid.UUID]) -> Dict[str, Project]:
    ids = list(project_ids)
    if not ids:
        return {}
    rows = (
        db.query(Project)
        .filter(Project.id.in_(ids), Project.deleted_at.is_(None))
        .all()
    )
    return {str(p.id): p for p in rows}


def resolve_job_label(
    db: Session,
    job_type: Optional[str],
    *,
    project_name: Optional[str] = None,
    projects_by_id: Optional[Dict[str, Project]] = None,
) -> Tuple[str, Optional[str]]:
    """Return (job_name, project_name) for display."""
    if not job_type:
        return project_name or "Unknown", project_name

    predefined = PREDEFINED_JOBS_DICT.get(job_type)
    if predefined:
        return predefined, project_name

    if _is_valid_uuid(job_type):
        project = None
        if projects_by_id is not None:
            project = projects_by_id.get(str(job_type))
        if project is None:
            try:
                project = (
                    db.query(Project)
                    .filter(Project.id == uuid.UUID(str(job_type)), Project.deleted_at.is_(None))
                    .first()
                )
            except (ValueError, TypeError):
                project = None
        if project is not None:
            resolved_name = format_project_job_label(project)
            return resolved_name, getattr(project, "name", None) or resolved_name

    if project_name:
        return project_name, project_name

    return "Unknown", project_name

File: frontend/src/lib/attendanceJobLabels.ts
import {
  formatJobPickerLine,
  getPredefinedJob,
} from '@/constants/predefinedJobs';

export type JobLabelSource = {
  id: string;
  name: string;
  code?: string | null;
};

export function shouldIgnoreApiJobName(jobName?: string | null): boolean {
  return !jobName || jobName.trim() === '' || jobName === 'Unknown';
}

export function parseJobTypeFromReasonText(reason?: string | null): string | null {
  if (!reason?.startsWith('JOB_TYPE:')) return null;
  const marker = reason.split('|')[0] ?? '';
  const jobType = marker.replace('JOB_TYPE:', '').trim();
  return jobType || null;
}

export function resolveJobLabelFromCache(
  jobTypeId: string | null | undefined,
  options: {
    apiJobName?: string | null;
    projectsById?: Record<string, JobLabelSource>;
    projectFromFetch?: JobLabelSource | null;
    shiftProject?: JobLabelSource | null;
  } = {},
): string | null {
  if (options.apiJobName && !shouldIgnoreApiJobName(options.apiJobName)) {
    return options.apiJobName;
  }

  if (!jobTypeId) return null;

  const predefined = getPredefinedJob(jobTypeId);
  if (predefined) return formatJobPickerLine(predefined);

  if (options.shiftProject) return formatJobPickerLine(options.shiftProject);

  const fromMap = options.projectsById?.[jobTypeId];
  if (fromMap) return formatJobPickerLine(fromMap);

  if (options.projectFromFetch) return formatJobPickerLine(options.projectFromFetch);

  return null;
}

export function resolveAttendanceEventJobLabel(
  event: {
    shift_id?: string | null;
    job_name?: string | null;
    project_name?: string | null;
    job_type?: string | null;
  },
  jobOptions: JobLabelSource[],
): string {
  if (event.shift_id) {
    return event.project_name || event.job_name || 'No Project';
  }

  const fromOptions = event.job_type
    ? jobOptions.find((j) => j.id === event.job_type)?.name
    : undefined;

  if (event.job_name && !shouldIgnoreApiJobName(event.job_name)) {
    return event.job_name;
  }
  if (event.project_name) return event.project_name;
  if (fromOptions) return fromOptions;
  if (event.job_type) return 'Unknown';
  return 'No Project';
}

export function resolveDirectAttendanceJobLabel(
  attendance: {
  shift_id?: string | null;
  job_name?: string | null;
  project_name?: string | null;
  reason_text?: string | null;
  job_type?: string | null;
}, options: {
  shiftProject?: JobLabelSource | null;
  projectsById?: Record<string, JobLabelSource>;
} = {}): string | null {
  if (attendance.shift_id) {
    if (options.shiftProject) return formatJobPickerLine(options.shiftProject);
    if (attendance.project_name) return attendance.project_name;
    return attendance.job_name || 'Unknown Project';
  }

  const jobTypeId =
    attendance.job_type || parseJobTypeFromReasonText(attendance.reason_text);

  return (
    resolveJobLabelFromCache(jobTypeId, {
      apiJobName: attendance.job_name,
      projectsById: options.projectsById,
    }) ?? jobTypeId
  );
}
